File: IN1010/trix/go/dobbeltlenketliste/dobbellenke.go
package main

type Dobbellenke struct {
	start *node
}

type node struct {
	forrige *node
	neste *node
	data *any
}

func (d *Dobbellenke) SettInn(T *any) {
	if d.start == nil {
		d.start = &node{
			forrige: nil,
			neste: nil,
			data: T,
		}
	} else {
		n := d.start
		for n.neste != nil {
			n = n.neste
		}
		n.neste = &node{
			forrige: n,
			neste: nil,
			data: T,
		}
	}
}

func (d *Dobbellenke) Lengde() int {
	cnt := 0
	n := d.start
	for n != nil {
		n = n.neste
		cnt++
	}
	return cnt
}

func (d *Dobbellenke) Hent(i int) *any {
	cnt := 0
	n := d.start
	for cnt != i && n != nil {
		n = n.neste
		cnt++
	}
	if n == nil {
		return nil
	}
	return n.data
}

func (d *Dobbellenke) Slett(i int) *any {
	cnt := 0
	n := d.start
	for cnt != i && n != nil {
		n = n.neste
		cnt++
	}
	if n == nil {
		return nil
	}

	prev := n.forrige
	next := n.neste

	if prev != nil {
		prev.neste = next
	} else {
		d.start = next
	}

	if next != nil {
		next.forrige = prev
	}
	
	return n.data
}

File: IN1010/trix/07/dobbellenketliste/Dobbellenke.java
class Dobbellenke<T> {
    Node start = null;

    class Node {
        T data;
        Node neste;
        Node forrige;

        Node(T d, Node n, Node f) {
            this.data = d;
            this.neste = n;
            this.forrige = f;
        }
    }

    public void SettInn(T data) {
        Node ny = new Node(data, null, null);
        if (this.start == null) {
            this.start = ny;
            return;
        }
        Node tmp = this.start;
        while (tmp.neste != null) {
            tmp = tmp.neste;
        }
        tmp.neste = ny;
        ny.forrige = tmp;
    }

    public T Fjern(int i) {
        int cnt = 0;
        Node tmp = this.start;
        while (cnt != i) {
            tmp = tmp.neste;
            cnt++;
        }
        if (tmp == null) {
            return null;
        }
        Node ret = tmp;

        Node prev = tmp.forrige;
        Node next = tmp.neste;

        if (prev != null) {
            prev.neste = next;
        } else {
            this.start = next;
        }

        if (next != null) {
            next.forrige = prev;
        }
        return ret.data;
    }

    public int Lengde() {
        if (this.start == null) {
            return 0;
        }
        int cnt = 0;
        Node tmp = this.start;
        while (tmp != null) {
            tmp = tmp.neste;
            cnt++;
        }
        return cnt;
    }
}
